lib/handler/gzip.c

```diff
@@ -0,0 +1,184 @@
+        out_size += inbufs[i].len;
+
+    outbuf_index = self->outbufs.size++;
+    self->outbufs.entries[outbuf_index].base = h2o_mem_alloc_pool(&req->pool, out_size);
+    /* set out buffer for output data from zlib */
+    self->zstream.next_out = (unsigned char *) self->outbufs.entries[outbuf_index].base;
+    self->zstream.avail_out = out_size;
+
+    for (i = 0; i != inbufcnt; ++i) {
+        int ret, flush;
+
+        self->zstream.next_in = (unsigned char*) inbufs[i].base;
+        self->zstream.avail_in = inbufs[i].len;
+        
+        if (i == inbufcnt - 1) {
+            flush = is_final ? Z_FINISH : Z_SYNC_FLUSH;
+        } else {
+            flush = Z_NO_FLUSH;
+        }
+        ret = deflate(&self->zstream, flush);
+        if (ret != Z_OK && ret != Z_STREAM_END) {
+            /* Should log and a proper way */
+            assert(0);
+        }
+
+    }
+    self->outbufs.entries[outbuf_index].len = out_size - self->zstream.avail_out;
+    h2o_ostream_send_next(&self->super, req, &self->outbufs.entries[outbuf_index], 1, is_final);
+}
+
+static void on_setup_ostream(h2o_filter_t *self, h2o_req_t *req, h2o_ostream_t **slot)
+{
+    gzip_encoder_t *encoder;
+
+    /* do nothing if content-length is unknown */
+    if (req->res.content_length == SIZE_MAX)
+        goto Next;
+    /* RFC 2616 4.4 states that the following status codes (and response to a HEAD method) should not include message body */
+    if ((100 <= req->res.status && req->res.status <= 199) || req->res.status == 204 || req->res.status == 304)
+        goto Next;
+    else if (h2o_memis(req->input.method.base, req->input.method.len, H2O_STRLIT("HEAD")))
+        goto Next;
+    /* we cannot handle certain responses (like 101 switching protocols) */
+    if (req->res.status != 200) {
+        req->http1_is_persistent = 0;
+        goto Next;
+    }
+    /* skip if transfer-encoding header is being set (for now) */
+    if (h2o_find_header(&req->res.headers, H2O_TOKEN_TRANSFER_ENCODING, -1) != -1)
+        goto Next;
+    /* skip if content-encoding header is being set */
+    if (h2o_find_header(&req->res.headers, H2O_TOKEN_CONTENT_ENCODING, -1) != -1)
+        goto Next;
+    /* 
+     * NOTE: I think there should be some test on MIME type, since we have no need to enable gzip on
+     *       those multimedia files or something like that.
+     */
+
+    /* set content-encoding header */
+    h2o_add_header(&req->pool, &req->res.headers, H2O_TOKEN_CONTENT_ENCODING, H2O_STRLIT("gzip"));
+    /* 
+     * For now, I need to change content_length to SIZE_MAX so I can enable chunked filter.
+     * I need chunked filter because I cannot know exactly how many times the "send_gzip()"
+     * function above will called at this point.
+     */
+    req->res.content_length = SIZE_MAX;
+
+    /* setup filter */
+    encoder = (void *)h2o_add_ostream(req, sizeof(gzip_encoder_t), slot);
+    encoder->super.do_send = send_gzip;
+    slot = &encoder->super.next;
+    
+    /*
```

Comment:
IMO there is no need to tweak window bits or memory level.  There is no need to define a custom callback for zalloc and zfree (the callbacks should be left NULL so that zlib would use the default ones).

All you need to do is delay the call to `deflateInit` (you do not need the flexibility provided by `deflateInit2`) until the first octet is received in `send_gzip`, and if `is_final` flag is set, call `deflateEnd` right after calling `deflate`.

If we implement the feature as such, the number of deflate contexts would be one-per-thread for small-sized responses, meaning that the impact on memory footprint would be minimal (and that is why you do not need to tweak window size or memory level for small responses).  It can also be said that allocating a fixed size of memory for every deflate context will actually be faster than allocating different sizes of blocks depending on the size of the response considering how `malloc` is implemented (in short, they have some kind of pool).  And in case the size of the content is large, window size and memory level cannot (should not) be reduced.  Therefore, there is no need to _tune_ the values.